src/transformers/models/blt_wip/modeling_blt_modular.py

```diff
@@ -0,0 +1,1249 @@
+
+from ...modeling_utils import ALL_ATTENTION_FUNCTIONS, PreTrainedModel
+from .configuration_blt import (
+    BLTConfig,
+    PatchingModeEnum,
+)
+
+from ..mllama.modeling_mllama import MllamaTextRMSNorm, MllamaTextMLP, MllamaTextCrossAttention, MllamaRotaryEmbedding, MllamaTextSelfAttention, MllamaSelfAttentionDecoderLayer, eager_attention_forward, repeat_kv
+
+if is_torch_flex_attn_available():
+    from torch.nn.attention.flex_attention import BlockMask
+
+    from ...integrations.flex_attention import make_flex_block_causal_mask
+
+
+logger = logging.get_logger(__name__)
+
+# Copied from transformers.models.mllama.modeling_mllama.MllamaTextMLP
+class BLTMLP(MllamaTextMLP):
+    pass
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    # TODO: not exactly equivalent to other transformers implementations,, need feedback
+    # Extract first head_dim//2 elements which correspond to the unique frequencies
+    # This matches the original BLT approach which uses head_dim//2 frequency pairs
+    head_dim = q.shape[-1]
+    cos_freqs = cos[..., :head_dim//2]  # [B, S, D/2]
+    sin_freqs = sin[..., :head_dim//2]  # [B, S, D/2]
+    
+    # Expand cos/sin to match query/key tensor format [B, H, S, D/2]
+    cos_freqs = cos_freqs.unsqueeze(1).expand(-1, q.shape[1], -1, -1)  # [B, 1, S, D/2] -> [B, H, S, D/2]
+    sin_freqs = sin_freqs.unsqueeze(1).expand(-1, q.shape[1], -1, -1)  # [B, 1, S, D/2] -> [B, H, S, D/2]
+    
+    # Split q and k into pairs for rotation: (d0, d1), (d2, d3), ...
+    q_pairs = q.view(*q.shape[:-1], head_dim//2, 2)  # [B, H, S, D/2, 2]
+    k_pairs = k.view(*k.shape[:-1], head_dim//2, 2)  # [B, H, S, D/2, 2]
+    
+    # Extract real and i parts
+    q_real, q_imag = q_pairs[..., 0], q_pairs[..., 1]  # [B, H, S, D/2]
+    k_real, k_imag = k_pairs[..., 0], k_pairs[..., 1]  # [B, H, S, D/2]
+    
+    # Apply rotation: [real', imag'] = [cos*real - sin*imag, sin*real + cos*imag]
+    q_real_rot = cos_freqs * q_real - sin_freqs * q_imag
+    q_imag_rot = sin_freqs * q_real + cos_freqs * q_imag
+    k_real_rot = cos_freqs * k_real - sin_freqs * k_imag
+    k_imag_rot = sin_freqs * k_real + cos_freqs * k_imag
+    
+    # Recombine pairs and reshape back to original format
+    q_rot = torch.stack([q_real_rot, q_imag_rot], dim=-1).view(*q.shape)  # [B, H, S, D]
+    k_rot = torch.stack([k_real_rot, k_imag_rot], dim=-1).view(*k.shape)  # [B, H, S, D]
+    
+    return q_rot.type_as(q), k_rot.type_as(k)
+
+# Copied from transformers.models.llama.modeling_llama.LlamaRMSNorm with Llama->MllamaText
+class BLTRMSNorm(MllamaTextRMSNorm):
+    pass
+
+
+class BLTTransformerLayer(MllamaSelfAttentionDecoderLayer):
+    pass
+
+
+# Copied from transformers.models.mllama.modeling_mllama.MllamaTextSelfAttention with MllamaText->BLT
+class BLTSelfAttention(MllamaTextSelfAttention):
+    pass
+
+
+def check_non_zero_after_zero(tensor):
```

Comment:
up for debate (not up to date with best practices with modular), but wondering if we shouldn't get all helper methods to be at the top of the file instead of across the file in between modules

Reply:
let's check how it's done in other modular files to see how we would want to approach it